src/Mod/Sketcher/Gui/SnapManager.h

```diff
@@ -0,0 +1,122 @@
+ ***************************************************************************/
+
+#ifndef SKETCHERGUI_SnapManager_H
+#define SKETCHERGUI_SnapManager_H
+
+
+#include <App/Application.h>
+
+
+namespace SketcherGui {
+
+class ViewProviderSketch;
+
+
+class ViewProviderSketchSnapAttorney {
+private:
+
+    static inline int getPreselectPoint(const ViewProviderSketch& vp);
+    static inline int getPreselectCross(const ViewProviderSketch& vp);
+    static inline int getPreselectCurve(const ViewProviderSketch& vp);
+
+    friend class SnapManager;
+};
+
+/* This class is used to manage the overriding of mouse pointer coordinates in Sketcher
+*  (in Edit-Mode) depending on the situation. Those situations are in priority order :
+*  1 - Snap at angle: For tools like Slot, Arc, Line, Ellipse, this enables to constrain the angle at steps of 5� (or customized angle).
+*  This is useful to make features at a certain angle (45� for example)
+*  2 - Snap to object: This snaps the mouse pointer onto objects.
+*  3 - Snap to grid: This snaps the mouse pointer on the grid.
+*/
+class SnapManager
+{
+
+    /** @brief      Class for monitoring changes in parameters affecting Snapping
+    *  @details
+    *
+    * This nested class is a helper responsible for attaching to the parameters relevant for
+    * SnapManager, initialising the SnapManager to the current configuration
+    * and handle in real time any change to their values.
+    */
+    class ParameterObserver : public ParameterGrp::ObserverType
+    {
+    public:
+        explicit ParameterObserver(SnapManager& client);
+        ~ParameterObserver() override;
+
+        void subscribeToParameters();
+
+        void unsubscribeToParameters();
+
+        /** Observer for parameter group. */
+        void OnChange(Base::Subject<const char*>& rCaller, const char* sReason) override;
+
+    private:
+        void initParameters();
+        void updateSnapToObjectParameter(const std::string& parametername);
+        void updateSnapToGridParameter(const std::string& parametername);
+        void updateSnapAngleParameter(const std::string& parametername);
+
+        static ParameterGrp::handle getParameterGrpHandle();
+
+    private:
+        std::map<std::string, std::function<void(const std::string&)>> str2updatefunction;
+        SnapManager& client;
+    };
+
+public:
+    explicit SnapManager(ViewProviderSketch &vp);
+    ~SnapManager();
+
+    bool snap(double& x, double& y);
+    bool snapAtAngle(double& x, double& y);
+    bool snapToObject(double& x, double& y);
+    bool snapToGrid(double& x, double& y);
+
+    bool snapToLineMiddle(Base::Vector3d& pointToOverride, const Part::GeomLineSegment* line);
+    bool snapToArcMiddle(Base::Vector3d& pointToOverride, const Part::GeomArcOfCircle* arc);
+
+    bool angleSnapEnabled;
+    Base::Vector2d referencePoint;
+    double snapAngle;
+    double lastMouseAngle;
+
+    bool snapToObjectsToggle;
+    bool snapToGridToggle;
+

```

Comment:
These last two "...Toggle". This Toggle is unclear and perhaps inconsistence with "angleSnapEnabled".

Given that it is not the state, but an account of the user preferences/tool requests, perhaps they could be "...Requested". This would also probably make the snap preference function easier to comprehend.